### PR TITLE
Initial Keccak256 hash implementation for circuit

### PR DESCRIPTION
**Pull request by ruslangm**

### CODE REVIEW

**Comment by ruslangm** [comment]
@lightning-li

File: std/hash/keccak256/keccak_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package keccak
+
+import (
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/test"
+	"golang.org/x/crypto/sha3"
```
Comment (ZhAnGeek) [inline std/hash/keccak256/keccak_test.go]: we better use keccak from go-ethereum.

"github.com/ethereum/go-ethereum/crypto"

crypto.Keccak256()

File: std/hash/keccak256/keccak_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		api.AssertIsEqual(result[i], circuit.ExpectedResult[i])
+	}
+	return nil
+}
+
+func TestKeccak256(t *testing.T) {
+	seed := new(big.Int).SetInt64(123456)
+	seedBytes := seed.FillBytes(make([]byte, 32))
+
+	hash := sha3.NewLegacyKeccak256()
```
Comment (ZhAnGeek) [inline std/hash/keccak256/keccak_test.go]: as above.

File: std/hash/keccak256/keccak.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+func (h *Keccak256) Write(data ...frontend.Variable) {
+	bs := h.BlockSize()
+
+	in := make([]keccakf2.Xuint8, len(data))
+	for i := range data {
+		in[i] = h.uapi8.AsUint8(data[i])
+	}
+
+	for len(in) > 0 {
```
Comment (ZhAnGeek) [inline std/hash/keccak256/keccak.go:71]: maybe we should flush first? in case buf got values in it.

**Review by ZhAnGeek — APPROVED** [review]
mostly good, some problems remain. and do we need to test more using fuzz or something?

File: std/hash/keccak256/keccak.go (hunk 3 above)
Comment (ruslangm) [reply std/hash/keccak256/keccak.go:71]: We always do flush as soon as underlying buffer length equals to `BlockSize`, even if `buf` got values in it.
So in such case `d.len` will hold a value which is `!=0` and another `Write` will add required amount of bytes to `buffer` and `flush` operation will be performed.
I will think about how to do it more clear and accordingly to `go-ethereum` implementation.

**Comment by ruslangm** [comment]
@ZhAnGeek yep, agree about more testing (also using fuzz), will add it to the PR.

File: std/hash/keccak256/keccak.go (hunk 3 above)
Comment (ruslangm) [reply std/hash/keccak256/keccak.go:71]: And I will add more test cases for such scenarios.

File: std/hash/keccak256/keccak_test.go (hunk 1 above)
Comment (ruslangm) [reply std/hash/keccak256/keccak_test.go]: done

File: std/hash/keccak256/keccak_test.go (hunk 2 above)
Comment (ruslangm) [reply std/hash/keccak256/keccak_test.go]: done

**Comment by ZhAnGeek** [comment]
lgtm

**Pull request merged**